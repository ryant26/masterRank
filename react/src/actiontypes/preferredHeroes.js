export const ADD_HERO = 'preferredHero/ADD_HERO';