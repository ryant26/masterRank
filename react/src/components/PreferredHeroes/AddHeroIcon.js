import React from 'react';
import FontAwesome from 'react-fontawesome';

let AddHeroIcon = () => {
  return (
    <div className="PreferredHeroes-AddHeroIcon flex justify-center align-center">
      <FontAwesome name="plus"/>
    </div>
  );
};

export default AddHeroIcon;
