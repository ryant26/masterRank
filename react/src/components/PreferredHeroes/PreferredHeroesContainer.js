import React, { Component } from 'react';
import HeroList from './HeroList';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

class PreferredHerosContainer extends Component {
  render() {
    const { heroes = [] } = this.props;

    return (
      <HeroList heroes={heroes}/>
    );
  }
}

const mapStateToProps = (state) => {
  return {
    heroes: state.preferredHeroes
  };
};



PreferredHerosContainer.propTypes = {
  heroes: PropTypes.array.isRequired
};

export default connect(mapStateToProps)(PreferredHerosContainer);