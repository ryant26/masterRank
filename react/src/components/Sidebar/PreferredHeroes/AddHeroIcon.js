import React from 'react';
import FontAwesome from 'react-fontawesome';
import PropTypes from 'prop-types';

let AddHeroIcon = (props) => {
  return (
    <div onClick={props.onClick} className="PreferredHeroes-AddHeroIcon flex justify-center align-center">
      <FontAwesome name="plus"/>
    </div>
  );
};

AddHeroIcon.propTypes = {
    onClick: PropTypes.func
};

export default AddHeroIcon;
