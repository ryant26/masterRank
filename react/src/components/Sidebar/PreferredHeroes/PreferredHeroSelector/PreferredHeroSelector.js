import React, { Component } from 'react';
import HeroSelector from '../../../ContentArea/DashboardHome/HeroSelector/HeroSelector';
import PropTypes from 'prop-types';
import HeroSlot from '../HeroSlot/HeroSlot';

class PreferredHeroSelector extends Component {

    constructor(props) {
        super(props);
        this.state = {
            selectedHero: props.preferredHeroes[props.selectedHeroSlot - 1],
            heroForLastAction: '',
            lastActionTaken: ''
        };

        this.maxSlots = 5;
        this.setSelectedHero = this.setSelectedHero.bind(this);
        this.setSelectedSlot = this.setSelectedSlot.bind(this);
        this.setHeroPreference = this.setHeroPreference.bind(this);
        this.clearHeroPreference = this.clearHeroPreference.bind(this);
    }

    componentWillReceiveProps(nextProps) {
        this.setSelectedHero(this.props.preferredHeroes[nextProps.selectedHeroSlot -1]);
    }

    setSelectedHero(heroName) {
        this.setState({
            selectedHero: heroName
        });
    }

    clearHeroPreference() {
        this.setState({
            heroForLastAction: this.state.selectedHero,
            lastActionTaken: `has been removed`
        });

        this.props.clearHeroPreference();
    }

    setHeroPreference(heroName) {
        if(!this.props.preferredHeroes.includes(heroName)){
            this.props.setHeroPreference(heroName);

            this.setState({
                heroForLastAction: heroName,
                lastActionTaken: `saved as preferred hero #${this.props.selectedHeroSlot}`
            });

            this.setSelectedSlot(this.props.selectedHeroSlot + 1);
        }
    }

    setSelectedSlot(slot) {
        this.props.setSelectedHeroSlot(slot);
    }

    render() {
        let heroSlots = [];
        for (let i = 1; i <= this.maxSlots; i++) {
            heroSlots.push(
                <HeroSlot
                    key={i}
                    slotNumber={i}
                    selectedSlotNumber={this.props.selectedHeroSlot}
                    preferredHeroes={this.props.preferredHeroes}
                    onSlotSelected={this.setSelectedSlot}
                />
            );
        }

        return (
            <div className="PreferredHeroSelector flex flex-column">
                <div className="header flex justify-between align-center">
                    <div>
                        <h3>Preferred Heroes</h3>
                        <div className="sub-title">Choose up to five preferred heroes to play.</div>
                    </div>
                    <div className="flex justify-between hero-slots">
                        {heroSlots}
                    </div>
                </div>
                <div className="body">
                    <HeroSelector selectedHeroes={[this.state.selectedHero]}
                                  disabledHeroes={this.props.preferredHeroes.filter((hero) => hero !== this.state.selectedHero)}
                                  onHeroSelected={this.setHeroPreference}/></div>
                <div className="footer flex align-center">
                    <div className="last-action-taken">
                        <b>{this.state.heroForLastAction.charAt(0).toUpperCase() + this.state.heroForLastAction.slice(1)}</b>{this.state.lastActionTaken}
                    </div>
                    <div className="button-group flex">
                        {this.props.preferredHeroes[this.props.selectedHeroSlot - 1] ? (
                            <div className="flex align-center button-six" onClick={this.clearHeroPreference}>
                                <div className="button-content">
                                    REMOVE
                                </div>
                            </div>
                        ) : undefined
                        }
                        {this.props.preferredHeroes.length ? (
                            <div className="save-button button-primary flex align-center" onClick={this.props.closeModal}>
                                <div className="button-content">
                                    SAVE
                                </div>
                            </div>
                        ) : undefined}

                    </div>
                </div>
            </div>
        );
    }
}

PreferredHeroSelector.propTypes = {
    selectedHeroSlot: PropTypes.number.isRequired,
    preferredHeroes: PropTypes.array.isRequired,
    setSelectedHeroSlot: PropTypes.func.isRequired,
    setHeroPreference: PropTypes.func.isRequired,
    clearHeroPreference: PropTypes.func.isRequired,
    closeModal: PropTypes.func.isRequired
};

export default PreferredHeroSelector;
