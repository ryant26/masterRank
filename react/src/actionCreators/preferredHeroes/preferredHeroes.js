import * as PreferredHeroActionsTypes from 'actiontypes/preferredHeroes';

export const addHero = (hero, preference) => {
  return {
    type: PreferredHeroActionsTypes.ADD_HERO,
    hero,
    preference
  };
};

export const removeHero = (hero, preference) => {
    return {
        type: PreferredHeroActionsTypes.REMOVE_HERO,
        hero,
        preference
    };
};

export const updateHeroes = (heroes) => {
    return {
        type: PreferredHeroActionsTypes.UPDATE_HEROES,
        heroes
    };
};