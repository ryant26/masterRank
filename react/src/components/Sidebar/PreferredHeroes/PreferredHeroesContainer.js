import React, {Component} from 'react';
import {connect} from 'react-redux';
import HeroImage from 'components/Images/HeroImage/HeroImage';
import AddHeroIcon from 'components/Sidebar/PreferredHeroes/AddHeroIcon';
import Modal from 'components/Modal/Modal';
import PreferredHeroSelector from 'components/Sidebar/PreferredHeroes/PreferredHeroSelector/PreferredHeroSelector';
import PropTypes from 'prop-types';
import Model from 'model/model';

class PreferredHeroesContainer extends Component {

    constructor(props) {
        super(props);
        this.state = {
            modalOpen: false,
            selectedSlot: 1,
            pendingPreferredHeroes: [...props.heroes]
        };

        this.maxSlots = 5;
        this.openModal = this.openModal.bind(this);
        this.closeModal = this.closeModal.bind(this);
        this.setHeroPreference = this.setHeroPreference.bind(this);
        this.setSelectedSlot = this.setSelectedSlot.bind(this);
        this.clearHeroPreference = this.clearHeroPreference.bind(this);
        this.handleVisibilityEvent = this.handleVisibilityEvent.bind(this);
    }


    componentDidMount() {
        window.addEventListener('visibilitychange', this.handleVisibilityEvent);
    }

    componentWillReceiveProps(nextProps) {
        this.setState(() => {
            return {
                pendingPreferredHeroes: [...nextProps.heroes]
            };
        });
    }

    componentWillUnmount() {
        window.removeEventListener('visibilitychange', this.handleVisibilityEvent);
    }

    setHeroPreference(heroName) {
        // In-place operation
        this.state.pendingPreferredHeroes.splice(this.state.selectedSlot - 1, 1, heroName);

        // Propigate change
        this.setState((state) => {
            return {
                pendingPreferredHeroes: state.pendingPreferredHeroes
            };
        });
    }

    clearHeroPreference() {
        // In-place operation
        this.state.pendingPreferredHeroes.splice(this.state.selectedSlot - 1, 1);

        // Propigate change
        this.setState((state) => {
            return {
                pendingPreferredHeroes: state.pendingPreferredHeroes
            };
        });
    }


    setSelectedSlot(slot) {
        let maxAllowableSlot = this.state.pendingPreferredHeroes.length + 1;
        let newSlot = maxAllowableSlot;

        if(slot <= this.maxSlots) {

            if (slot <= maxAllowableSlot) {
                newSlot = slot;
            }

            this.setState({
                selectedSlot: newSlot
            });
        }
    }

    openModalWithSlot(slot) {
        return () => {
            this.setSelectedSlot(slot);
            this.openModal();
        };
    }


    openModal() {
        this.setState(() => {
            return {
                modalOpen: true
            };
        });
    }

    closeModal() {
        Model.updatePreferredHeroes(this.state.pendingPreferredHeroes);
        this.setState(() => {
            return {
                modalOpen: false
            };
        });
    }

    handleVisibilityEvent() {
        Model.updatePreferredHeroes(JSON.parse(window.localStorage.getItem('state')).preferredHeroes.heroes);
    }

    render() {
        let heroThumbnails = this.props.heroes.map((hero, i) => {
            return (
                <HeroImage key={hero} heroName={hero} onClick={this.openModalWithSlot(i + 1)}/>
            );
        });

        for (let i = heroThumbnails.length; i < 5; i++) {
            heroThumbnails.push(<AddHeroIcon key={i} onClick={this.openModalWithSlot(i + 1)}/>);
        }

        return (
            <div className="PreferredHeroes-HeroList sidebar-card flex flex-column">
                <div className="sidebar-title">Heroes I'll Play</div>
                <div className="flex justify-between">
                    {heroThumbnails}
                </div>
                <Modal closeModal={this.closeModal} modalOpen={this.state.modalOpen}>
                    <PreferredHeroSelector
                        setSelectedHeroSlot={this.setSelectedSlot}
                        setHeroPreference={this.setHeroPreference}
                        clearHeroPreference={this.clearHeroPreference}
                        preferredHeroes={this.state.pendingPreferredHeroes}
                        selectedHeroSlot={this.state.selectedSlot}
                        closeModal={this.closeModal}
                    />
                </Modal>
            </div>
        );
    }
}

PreferredHeroesContainer.propTypes = {
    heroes: PropTypes.array.isRequired,
};

const mapStateToProps = (state) => {
    return {
        heroes: state.preferredHeroes.heroes
    };
};

export default connect(mapStateToProps)(PreferredHeroesContainer);