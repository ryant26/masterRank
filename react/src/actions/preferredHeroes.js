import * as PreferredHeroActionsTypes from '../actiontypes/preferredHeroes';

export const addHero = (hero, preference) => {
  return {
    type: PreferredHeroActionsTypes.ADD_HERO,
    hero,
    preference
  };
};

export const setSelectedSlot = (slot) => {
    return {
        type: PreferredHeroActionsTypes.SET_SELECTED_SLOT,
        slot
    };
};