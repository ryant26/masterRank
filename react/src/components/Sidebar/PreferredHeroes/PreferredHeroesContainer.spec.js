import React from 'react';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';
import {createStore} from '../../../model/store';
import Model from '../../../model/model';
import {addHero} from '../../../actionCreators/preferredHeroes/preferredHeroes';
import AddHeroIcon from './AddHeroIcon';
import PreferredHeroSelector from './PreferredHeroSelector/PreferredHeroSelector';

const MockSocket = require('socket-io-mock');

import PreferredHeroesContainer from './PreferredHeroesContainer';

const mountWithHeroes = (heroes=[], store=createStore()) => {
    heroes.forEach((hero, i) => {
        store.dispatch(addHero(hero, i + 1));
    });

    return mount(
        <PreferredHeroesContainer store={store}/>
        , {context: {store}}
    );
};

describe('<PreferredHeroesContainer/>', () => {
    let wrapper;
    let store;

    beforeEach(() => {
        store = createStore();
        Model.initialize(new MockSocket(), store);
        wrapper = mountWithHeroes([], store);
    });

    it('Should render without exploding', () => {
        const PreferredHerosContainer = wrapper.find(PreferredHeroesContainer);
        expect(PreferredHerosContainer.length).toBeTruthy();
    });

    it('Should render in the correct format without any preferred heroes', () => {
        const component = renderer.create(
            <PreferredHeroesContainer store={store}/>
        );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('Should render in the correct format with preferred heroes', () => {
        store.dispatch(addHero('genji', 1));
        store.dispatch(addHero('tracer', 2));
        store.dispatch(addHero('winston', 3));
        store.dispatch(addHero('mei', 4));
        store.dispatch(addHero('phara', 5));

        const component = renderer.create(
            <PreferredHeroesContainer store={store}/>
        );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('Should open the modal when a slot is clicked', function() {
        // State and DOM should match
        expect(wrapper.find("PreferredHeroesContainer").instance().state.modalOpen).toBeFalsy();
        expect(wrapper.find(PreferredHeroSelector).length).toBeFalsy();

        let node = wrapper.find(AddHeroIcon).at(0);
        node.simulate('click');

        // State and DOM should match
        expect(wrapper.find("PreferredHeroesContainer").instance().state.modalOpen).toBeTruthy();
        expect(wrapper.find(PreferredHeroSelector).length).toBeTruthy();
    });

    it('Should update preferred heroes on modal close', function() {
        const heroName = 'doomfist';
        let realFunction = Model.updatePreferredHeroes;
        Model.updatePreferredHeroes = jest.fn();

        let node = wrapper.find(AddHeroIcon).at(0);
        node.simulate('click');

        wrapper.find('PreferredHeroesContainer').instance().setHeroPreference(heroName);

        expect(wrapper.find("PreferredHeroesContainer").instance().state.modalOpen).toBeTruthy();
        expect(wrapper.find(PreferredHeroSelector).length).toBeTruthy();

        expect(Model.updatePreferredHeroes).not.toHaveBeenCalledWith([heroName]);

        wrapper.find('.close-button').simulate('click');

        expect(Model.updatePreferredHeroes).toHaveBeenCalledWith([heroName]);
        Model.updatePreferredHeroes = realFunction;
    });

    describe('setHeroPreference', () => {
        it('Should update the hero for the currently selected slot', () => {
            const heroName = 'doomfist';
            expect(wrapper.find('PreferredHeroesContainer').instance().state.pendingPreferredHeroes).toEqual(expect.arrayContaining([]));
            wrapper.find('PreferredHeroesContainer').instance().setHeroPreference(heroName);
            expect(wrapper.find('PreferredHeroesContainer').instance().state.pendingPreferredHeroes).toEqual(expect.arrayContaining([heroName]));
        });
    });

    describe('clearHeroPreference', () => {
        it('Should update the hero for the currently selected slot', () => {
            wrapper = mountWithHeroes(['genji', 'tracer']);

            expect(wrapper.find('PreferredHeroesContainer').instance().state.pendingPreferredHeroes).toEqual(expect.arrayContaining(['genji', 'tracer']));
            wrapper.find('PreferredHeroesContainer').instance().clearHeroPreference();
            expect(wrapper.find('PreferredHeroesContainer').instance().state.pendingPreferredHeroes).toEqual(expect.arrayContaining(['tracer']));
        });
    });

    describe('setSelectedSlot', () => {
        it('Should allow any slot with a preferenced hero to be selected', () => {
            wrapper = mountWithHeroes(['genji', 'tracer']);

            expect(wrapper.find('PreferredHeroesContainer').instance().state.selectedSlot).toEqual(1);
            wrapper.find('PreferredHeroesContainer').instance().setSelectedSlot(2);
            expect(wrapper.find('PreferredHeroesContainer').instance().state.selectedSlot).toEqual(2);
        });

        it('Should allow default to the lowest unpreferenced slot when passed a higher slot', () => {
            wrapper = mountWithHeroes(['genji', 'tracer']);

            expect(wrapper.find('PreferredHeroesContainer').instance().state.selectedSlot).toEqual(1);
            wrapper.find('PreferredHeroesContainer').instance().setSelectedSlot(4);
            expect(wrapper.find('PreferredHeroesContainer').instance().state.selectedSlot).toEqual(3);
        });
    });
});