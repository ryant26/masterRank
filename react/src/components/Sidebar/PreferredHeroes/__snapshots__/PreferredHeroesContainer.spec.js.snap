// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PreferredHeroesContainer/> Should render in the correct format with preferred heroes 1`] = `
<div
  className="PreferredHeroes-HeroList sidebar-card flex flex-column"
>
  <div
    className="sidebar-title"
  >
    Heroes I'll Play
  </div>
  <div
    className="flex justify-between"
  >
    <div
      className=""
    >
      <img
        alt="genji icon"
        className="HeroImage"
        onClick={[Function]}
        src={
          Object {
            "process": [Function],
          }
        }
      />
    </div>
    <div
      className=""
    >
      <img
        alt="tracer icon"
        className="HeroImage"
        onClick={[Function]}
        src={
          Object {
            "process": [Function],
          }
        }
      />
    </div>
    <div
      className=""
    >
      <img
        alt="winston icon"
        className="HeroImage"
        onClick={[Function]}
        src={
          Object {
            "process": [Function],
          }
        }
      />
    </div>
    <div
      className=""
    >
      <img
        alt="mei icon"
        className="HeroImage"
        onClick={[Function]}
        src={
          Object {
            "process": [Function],
          }
        }
      />
    </div>
    <div
      className=""
    >
      <img
        alt="phara icon"
        className="HeroImage"
        onClick={[Function]}
        src={
          Object {
            "process": [Function],
          }
        }
      />
    </div>
  </div>
</div>
`;

exports[`<PreferredHeroesContainer/> Should render in the correct format without any preferred heroes 1`] = `
<div
  className="PreferredHeroes-HeroList sidebar-card flex flex-column"
>
  <div
    className="sidebar-title"
  >
    Heroes I'll Play
  </div>
  <div
    className="flex justify-between"
  >
    <div
      className="PreferredHeroes-AddHeroIcon flex justify-center align-center"
      onClick={[Function]}
    >
      <span
        aria-hidden={true}
        className="fa fa-plus"
      />
    </div>
    <div
      className="PreferredHeroes-AddHeroIcon flex justify-center align-center"
      onClick={[Function]}
    >
      <span
        aria-hidden={true}
        className="fa fa-plus"
      />
    </div>
    <div
      className="PreferredHeroes-AddHeroIcon flex justify-center align-center"
      onClick={[Function]}
    >
      <span
        aria-hidden={true}
        className="fa fa-plus"
      />
    </div>
    <div
      className="PreferredHeroes-AddHeroIcon flex justify-center align-center"
      onClick={[Function]}
    >
      <span
        aria-hidden={true}
        className="fa fa-plus"
      />
    </div>
    <div
      className="PreferredHeroes-AddHeroIcon flex justify-center align-center"
      onClick={[Function]}
    >
      <span
        aria-hidden={true}
        className="fa fa-plus"
      />
    </div>
  </div>
</div>
`;
